features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/voicemessages/timeline/VoiceMessagePresenter.kt

```diff
@@ -0,0 +1,152 @@
+import io.element.android.features.messages.impl.timeline.model.event.TimelineItemVoiceContent
+import io.element.android.libraries.architecture.Async
+import io.element.android.libraries.architecture.Presenter
+import io.element.android.libraries.architecture.runUpdatingState
+import io.element.android.libraries.di.RoomScope
+import io.element.android.libraries.matrix.api.media.MatrixMediaLoader
+import io.element.android.libraries.matrix.api.media.MediaFile
+import io.element.android.libraries.matrix.api.media.toFile
+import kotlinx.coroutines.launch
+
+@Module
+@ContributesTo(RoomScope::class)
+interface VoiceMessagePresenterModule {
+    @Binds
+    @IntoMap
+    @TimelineItemEventContentKey(TimelineItemVoiceContent::class)
+    fun bindVoiceMessagePresenterFactory(factory: VoiceMessagePresenter.Factory): TimelineItemPresenterFactory<*, *>
+}
+
+class VoiceMessagePresenter @AssistedInject constructor(
+    private val mediaLoader: MatrixMediaLoader,
+    voiceMessagePlayerFactory: VoiceMessagePlayer.Factory,
+    voiceMessageCacheFactory: VoiceMessageCache.Factory,
+    @Assisted private val content: TimelineItemVoiceContent,
+) : Presenter<VoiceMessageState> {
+
+    @AssistedFactory
+    fun interface Factory : TimelineItemPresenterFactory<TimelineItemVoiceContent, VoiceMessageState> {
+        override fun create(content: TimelineItemVoiceContent): VoiceMessagePresenter
+    }
+
+    private val voiceCache = voiceMessageCacheFactory.create(mxcUri = content.mediaSource.url)
+
+    private val player = voiceMessagePlayerFactory.create(
+        eventId = content.eventId,
+        mediaPath = voiceCache.cachePath
+    )
+
+    @Composable
+    override fun present(): VoiceMessageState {
+
+        val scope = rememberCoroutineScope()
+
+        val playerState by player.state.collectAsState(VoiceMessagePlayer.State(isPlaying = false, isMyMedia = false, currentPosition = 0L))
+        val mediaFile = remember { mutableStateOf<Async<MediaFile>>(Async.Uninitialized) }
+
+        val button by remember {
+            derivedStateOf {
+                when {
+                    content.eventId == null -> VoiceMessageState.Button.Disabled
+                    playerState.isPlaying -> VoiceMessageState.Button.Pause
+                    mediaFile.value is Async.Loading -> VoiceMessageState.Button.Downloading
+                    mediaFile.value is Async.Failure -> VoiceMessageState.Button.Retry
+                    else -> VoiceMessageState.Button.Play
+                }
+            }
+        }
+        val progress by remember {
+            derivedStateOf { if (playerState.isMyMedia) playerState.currentPosition / content.duration.toMillis().toFloat() else 0f }
+        }
+        val time by remember {
+            derivedStateOf {
+                val time = if (playerState.isMyMedia) playerState.currentPosition else content.duration.toMillis()
+                "%02d:%02d".format(time.msToMins(), time.msToSecs())
```

Comment:
FYI we have a utility for this formatting: `Duration.formatShort()`.

Reply:
Yeah now we do :) switched to it.